### ISSUE TITLE
AutoML solutions review

### ISSUE THREAD

**Issue by martins0n** [tinkoff-ai/etna#804]
### Preposition:
We have multiple models and feature generators to use.
We don't know what model  or features to choose from.
We have limited cpu and time resources.
We have only target metric we want to optimize.
We have multiple nodes to run on.

### Aim:
Library level instrument to find out the best possible forecasting approach in the defined above constraints.

## Checklist:
- time series support
   - Yea/No
   - How it works?
- distributed
   - Yea/No
   - How is implemented and what is used for job scheduling-running
   - Could it work in detach mode or workers should have permanent connection with master node?
- feature engineering/selection
   - Yea/No
   - How feature space is generated?
   - We optimize full global space of chosen features, models, hyperparameters at once, or make it in waterfall maner: choose features with any method and then turning model-parameters.
- fault tolerance
   - What happenes if some trials are raising errors for example
   - Could we restart job from stopped point or we should start optimization from the beginning
- as a service
   - Can we expose only api and run jobs via requests.

##  Table:

| library | time series support | distributed | feature engineering/selection | fault tolerance |  as a service | comments |
| ----------- |  --- | --- | --- | --- |  --- | --- | 
| Flaml |  ✅[notebook](https://github.com/microsoft/FLAML/blob/89d9c4426ec092ec6f29f5d22ccb8ece461a8bc0/notebook/automl_time_series_forecast.ipynb) <br> 1. regression models with lags(hcrystalball backed) + ts specific models like prophet and sarima. </br> 2. No multiple series support |  ray as backend | <div> Feature selection ♾️ [issue](https://github.com/microsoft/FLAML/issues/258)  </div> <div> Hardcoded feature generation [ref](https://github.com/microsoft/FLAML/blob/d649fefa6b804b75ccdfc8acbfd9f2b98f849924/flaml/data.py#L251) </div>  | `flaml.tune` has parameter `max_failure` but it doesn't work as expected i guess | 🚫  | 1.[zero-shot automl](https://microsoft.github.io/FLAML/docs/Use-Cases/Zero-Shot-AutoML): features for metalearner: Dataset,NumberOfInstances,NumberOfFeatures,NumberOfClasses,PercentageOfNumericFeatures <br> 2.[Additional algorithms](https://github.com/microsoft/FLAML/tree/main/flaml/tune) for HPO </br> <br> 3. Predefined `low_cost_init_value` in `search_space()` of models for cost-efficient search. </br>
| h2o | 🚫 There is no classical ts support. You can work with timeseries only via tabular approach | Spark, local threads | There is no special auto feature engineering in free version.    | [ref](https://docs.h2o.ai/h2o/latest-stable/h2o-docs/grid-search.html#fault-tolerant-grid-search) Resume support after failing and caching.| Rest API and h2o flow |<div> 1. Predefined grid for hyps search </div> <div> 2. H20 is more about black box service. You can't extend it easily: it uses java for example but it should work the same way whatever the data size: if you have hadoop cluster or spark it wil automatically compute job there -- all infrastructure, datasources, algorithms are implemented already for distributed computing.</div>    |
| evalml | ✅ [example](https://evalml.alteryx.com/en/stable/user_guide/timeseries.html) <br> 1. [No multiple series support](https://github.com/alteryx/evalml/issues/1506) </br> <br> 2. Arima, prophet and regression models </br> | dask backed  or multiprocessing [ref](https://evalml.alteryx.com/en/stable/user_guide/automl.html#Parallel-AutoML)| You should preprocess data by hand with special transforms like `TimeSeriesImputer` for example. But after turning start they have [`TimeSeriesFeaturizer`](https://evalml.alteryx.com/en/stable/autoapi/evalml/pipelines/components/index.html#evalml.pipelines.components.TimeSeriesFeaturizer) for feature generation - it crates lags and select important lags via correlations. Full dag of [generation](https://evalml.alteryx.com/en/stable/_images/user_guide_timeseries_51_0.svg) | [callbacks for error handling](https://evalml.alteryx.com/en/stable/user_guide/automl.html#Error-Callback) | 🚫| <br> [DAG of feature generation pipeline](https://evalml.alteryx.com/en/stable/user_guide/objectives.html) </br> |
| PyCaret |  ✅ [models](https://github.com/pycaret/pycaret/blob/c7cb75b9b2628d6f9ecf7f3fe55b006ba2278a80/pycaret/time_series/forecasting/oop.py#L1902)  - predefined pipelines with preprocessing and feature generation <div>1.based on sktime</div> <div> 2. No multiple series support </div> | via fugue( dask + spark) | you can use flags `setup` to make preprocessing, feature selection | <div> 1. There is fitted transform cache </div> <div> 2. It seems there is no parameters for error handling settings </div> | 🚫 | <div> 1.After best model selection you can tune hyperparameters via `model.tune_model`. It seems some models have predefined grid in `tune_grid` attribute but in most cases you have too specify grid via `custom_grid` parameter </div><div>2. Beautiful docs via git book and cheatsheet</div> |
| LightAutoML | 🚫 | multiprocessing | <div> 1. There are predefined pipelines  - [example](https://github.com/sb-ai-lab/LightAutoML/blob/53cd4a7634099eb8815dd5d9f0dda5eaae51c3c2/lightautoml/pipelines/features/lgb_pipeline.py)</div> <div> 2. You can customize pipelines with feature selection steps or add new features via library classes </div> | There is no special flags or memoization | 🚫 | |
| MLJAR | 🚫 | multiprocessing | <div> 1. [Golden Features](https://supervised.mljar.com/features/golden_features/) with [feature selection](https://supervised.mljar.com/features/features_selection/) </div> |  <div> 1.[Progress tracking with artifacts with memoization. You can continue progress after restart if `results_path` specified ](https://github.com/mljar/mljar-supervised/blob/31e95852ccdd1a0338292965474d8ad1f9c36885/supervised/base_automl.py#L658) </div> | 🚫 | <div> 1. Good documentation and README.md </div> <div> 2. 4 Mods of training: Explain, Optuna, Compete, Perform </div> |
| TPOT | 🚫 | dask backed | feature construction from predefined configs and advanced feature selection modes  | warm_start  and memoization - you can continue progress after ctrl+c | 🚫  | <div> 1. [Default pipeline configurations](https://github.com/EpistasisLab/tpot/tree/master/tpot/config) </div> <div> 2. Good docs </div> |
| microsoft/nni | 🚫 | Multiple ways to distribute tasks | There is some classes for feature selection but they are not integrated in general pipeline, they live alone | You can resume experiment. All artifacts are persist to disk | there is web interface for experiment monitoring | It's more about experiment tracking with multiple extensions like HPO. It could be considered as open-source alternative to wandb. There is no any classical automl dependencies but you can build own automl based on nni infrastructure. | 
| [auto-sklearn](https://github.com/automl/auto-sklearn) | 🚫 | dask backed | BO on predefined feature+model space: start point is determined by knowledge of performance on OpenML dataset | there is no checkpointing or reload | 🚫 |  <div> 1. You can add custom pipelines or limit search space on feature model granularity </div>  <div> 2. This library more about optimal search algorithms than customisable and extensibility: the team developed SMAC algorithm -- so it should be interesting to read there papers first of all </div> <div> 3. [Good docs](https://automl.github.io/auto-sklearn/master/index.html) |
| [autogluon](https://github.com/awslabs/autogluon) | ✅ Multiple series support based on sktime and gluonts models | Ray backed (only for hpo) | In case of TS - no. For tabular data: for numerical -- there is no extra generation or preprocessing, for categorical, datetime types there are extra [transformations](https://auto.gluon.ai/stable/tutorials/tabular_prediction/tabular-feature-engineering.html) | `TimeSeriesPredictor` doesn't have any checkpointing or explicit error handlers | 🚫 | <div> 1. Good integration with gluon-ts </div> |
| FEDOT | ✅ multi-target or one-dimensional time-series. It seems there is no multiple series support | multiprocessing or proprietary cloud | Dynamic graph optimisation via genetic algorithms. Much in common with autosktime and TPOT. | Results caching in `history_folder`. But it seems there is no support of fit continuation | 🚫 | <div> 1. DAG and genetic algorithms </div> <div> 2. Bad documentation </div> | 

### Notes
- baseline models with heuristics, common models and ensembling at the end 

### Additional material
- https://github.com/hibayesian/awesome-automl-papers
